File: lis.cpp
```cpp
// This program finds the length of the longest increasing subsequence in an array.
// By: Jonathan Mitelman
#include <iostream>
#include <vector>
#include <algorithm>

int longest_increasing_subsequence(const std::vector<int>& arr) {
    int n = arr.size();
    if (n == 0) {
        return 0;
    }

    std::vector<int> dp(n, 1);

    for (int i = 1; i < n; ++i) {
        for (int j = 0; j < i; ++j) {
            if (arr[i] > arr[j]) {
                dp[i] = std::max(dp[i], dp[j] + 1);
            }
        }
    }

    return *std::max_element(dp.begin(), dp.end());
}

int main() {
    std::vector<int> arr1 = {10, 22, 9, 33, 21, 50, 41, 60, 80};
    std::vector<int> arr2 = {3, 10, 2, 1, 20};
    std::vector<int> arr3 = {50, 3, 10, 7, 40, 80};
    std::vector<int> arr4 = {10, 9, 8, 7, 6, 5};
    std::vector<int> arr5 = {};
    std::vector<int> arr6 = {1};

    std::cout << "LIS of {10, 22, 9, 33, 21, 50, 41, 60, 80}: " << longest_increasing_subsequence(arr1) << std::endl;
    std::cout << "LIS of {3, 10, 2, 1, 20}: " << longest_increasing_subsequence(arr2) << std::endl;
    std::cout << "LIS of {50, 3, 10, 7, 40, 80}: " << longest_increasing_subsequence(arr3) << std::endl;
    std::cout << "LIS of {10, 9, 8, 7, 6, 5}: " << longest_increasing_subsequence(arr4) << std::endl;
    std::cout << "LIS of {}: " << longest_increasing_subsequence(arr5) << std::endl;
    std::cout << "LIS of {1}: " << longest_increasing_subsequence(arr6) << std::endl;

    return 0;
}
```

File: lis.py
```python
"""Finds the longest increasing subsequence (LIS) of a given array.
By: Jonathan Mitelman"""
def longest_increasing_subsequence(arr):
    n = len(arr)
    if n == 0:
        return []

    tails = []
    predecessors = {}

    for number in arr:
        if tails and number <= tails[-1]:
            l, r = 0, len(tails) - 1
            while l <= r:
                mid = (l + r) // 2
                if tails[mid] < number:
                    l = mid + 1
                else:
                    r = mid - 1

            if tails[l] != number:
                tails[l] = number
                if l > 0:
                    predecessors[number] = tails[l - 1]

        else:
            if tails:
                predecessors[number] = tails[-1]
            tails.append(number)

    longest_subsequence = []
    current = tails[-1]
    while current is not None:
        longest_subsequence.insert(0, current)
        current = predecessors.get(current, None)

    return longest_subsequence


if __name__ == "__main__":
    test_array = [10, 9, 2, 5, 3, 7, 101, 18]
    lis = longest_increasing_subsequence(test_array)
    print(f"Original array: {test_array}")
    print(f"Longest Increasing Subsequence: {lis}")

    test_array2 = [0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15]
    lis2 = longest_increasing_subsequence(test_array2)
    print(f"Original array: {test_array2}")
    print(f"Longest Increasing Subsequence: {lis2}")

    test_array3 = []
    lis3 = longest_increasing_subsequence(test_array3)
    print(f"Original array: {test_array3}")
    print(f"Longest Increasing Subsequence: {lis3}")

    test_array4 = [1,3,6,7,9,4,10,5,6]
    lis4 = longest_increasing_subsequence(test_array4)
    print(f"Original array: {test_array4}")
    print(f"Longest Increasing Subsequence: {lis4}")
```